### DISCUSSION TITLE
Iterating liquidity from the current tick

### DISCUSSION THREAD

**Discussion by rracinskij** [atiselsts/uniswap-v3-liquidity-math#9] [Q&A] [answered]
Hello, I'd greatly appreciate your help on `subgraph-liquidity-range-example.py` script.

It iterates currently from zero (line 108) and the comment in the line 107 states that it should start from the total liquidity if iterating from the current tick. However, I am not sure how to change the code in order to do that.

Another question is whether the script takes into the account the problem with the unclaimed fees [https://medium.com/coinmonks/all-your-uniswap-v3-liquidity-farming-calculations-are-dead-wrong-heres-why-20bd47f55d69](url)?

**Comment by atiselsts** [comment, marked as answer]
Hi,

The active liquidity in the pool is read simply by accesing `pool["liquidity"]`.

About the problem mentioned in the article - the TVL problem shows up in the aggregate token values maintained by the subgraph. My scripts does not use those values, they use liquidity values, calculate token amounts from that, and sum over multiple ticks or positions. They should not be affected by the problem.

**Reply by rracinskij** [reply]
Thank you. I'm querying tick liquidity for a given range around the current range and get negative liquidity for some of the ticks. Is it possible or is it rather a bug in my code?
```
tick_curr = int(pool["tick"])
tick_spacing = fee_tier_to_tick_spacing(int(pool["feeTier"]))
tick_bottom = math.floor(tick_curr / tick_spacing) * tick_spacing
tick_top = tick_bottom + tick_spacing

tick_min = tick_bottom - tick_spacing * 50 # get liquidity for 50 ticks around the current one
tick_max = tick_bottom + tick_spacing * 50
```

**Reply by atiselsts** [reply]
Negative liquidity must be a bug somewhere.

**Reply by rracinskij** [reply]
But it is returned by the Graph, like `{'tickIdx': '200970', 'liquidityNet': '-207451718546852'}`, can't figure out why. It is a Polygon pool 0x45dda9cb7c25131df268515131f647d726f50608.

**Reply by atiselsts** [reply]
`liquidityNet` is the delta that you must subtract from or add to the current liquidity to get the next range's liquidity.

**Reply by rracinskij** [reply]
As an example - the pool query returns current tick equal to `202018` and the pool liquidity value of `2157315830892103776`. Then the `bottom_tick` is `202010` and the `top_tick` is `202020`.

Tick query returns these liquidity deltas: 

`202010: -43569232006628558, 202020: -11805592301812422.`

I am not sure how to interpret these deltas to calculate the liquidities of the adjacent tick ranges starting from the current tick range.

**Reply by atiselsts** [reply]
The meaning of `liquidityNet` is described in the whitepaper.

In your example:
[202010, 202020] range liquidity is 2157315830892103776

[202020, 202030] range liquidity is 2157315830892103776 + -11805592301812422 = 2145510238590291354

[202000, 202010] range liquidity is 2157315830892103776 - -43569232006628558 = 2200885062898732334 

In the final calculation you need to subtract because price (ticks) are decreasing when moving to this range.

**Reply by rracinskij** [reply]
Thank you. The idea was clear, but I was not sure which tick liquidity should be added to the current pool liquidity. I tested it on pool data and the numbers match.
